tests/apps/stateapp/app.go

```diff
@@ -0,0 +1,226 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"time"
+	"strings"
+
+	"github.com/gorilla/mux"
+)
+
+const appPort = 3000
+const stateURL = "http://localhost:3500/v1.0/state"
+
+// appState represents a state in this app.
+type appState struct {
+	Data   string `json:"data,omitempty"`
+}
+
+// daprState represents a state in Dapr.
+type daprState struct {
+	Key    string   `json:"key,omitempty"`
+	Value  *appState `json:"value,omitempty"`
+}
+
+// requestResponse represents a request or response for the APIs in this app.
+type requestResponse struct {
+	StartTime int           `json:"start_time,omitempty"`
+	EndTime   int           `json:"end_time,omitempty"`
+	States    []daprState   `json:"states,omitempty"`
+	Message   string        `json:"message,omitempty"`
+}
+
+// indexHandler is the handler for root path
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("indexHandler is called")
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func save(states []daprState) error {
+	log.Printf("Processing save request for %d entries.", len(states))
+
+	jsonValue, err := json.Marshal(states)
+	if err != nil {
+		log.Print("Could save states in Dapr: %s", err.Error())
+		return err
+	}
+
+	log.Printf("Posting state to %s with '%s'", stateURL, jsonValue)
+	_, err = http.Post(stateURL, "application/json", bytes.NewBuffer(jsonValue))
+
+	return err
+}
+
+func get(key string) (*appState, error) {
+	log.Printf("Processing get request for %s.", key)
+	url := strings.Join([]string { stateURL, key }, "/")
+	log.Printf("Fetching state from %s", url)
+	res, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("Could not get value for key %s from Dapr: %s", key, err.Error())
+	}
+
+	body, err := ioutil.ReadAll(res.Body)
+    if err != nil {
+        return nil, fmt.Errorf("Could not load value for key %s from Dapr: %s", key, err.Error())
+	}
+	
+	log.Printf("Found state for key %s: %s", key, body)
+
+	var state = new(appState)
+	if len(body) == 0 {
+		return nil, nil
+	}
+
+	// a key not found in Dapr will return 200 but an empty response.
+	err = json.Unmarshal([]byte(body), &state)
```

Comment:
we can define `state` variable before line 87 instead of creating the instance of appState in the heap.

```
var state appState
```

Reply:
Sorry I'm wrong both are going to heap.